llvm/lib/Transforms/Utils/CanonicalizeFreezeInLoops.cpp

```diff
@@ -151,11 +151,18 @@ bool CanonicalizeFreezeInLoopsImpl::run() {
       }
     }
 
+    bool Exist = false;
     auto Visit = [&](User *U) {
       if (auto *FI = dyn_cast<FreezeInst>(U)) {
-        LLVM_DEBUG(dbgs() << "canonfr: found: " << *FI << "\n");
-        Info.FI = FI;
-        Candidates.push_back(Info);
+        for (const auto &Candidate : Candidates) {
```

Comment:
How about instead changing Candidates to a SetVector as suggested by @nikic at the issue? I suppose it would require specializing `DenseMapInfo` for `FrozenIndPHIInfo`. See https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/MemoryLocation.h#L330 for an example

Reply:
The fix has been refactored by using SmallSetVector.
Thanks for the example that really helps a lot!